In [1]:
from sklearn.datasets import load_boston
import pandas as pd
import numpy as np

In [2]:
RANDOM_STATE = 42

In [3]:
dataset = load_boston()
X = pd.DataFrame(dataset.data)
X.columns = dataset.feature_names
y = dataset.target

1. Разделите выборку на обучающую и тестовую в отношении 80% / 20%.

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((404, 13), (102, 13), (404,), (102,))

2. Обучите стандартную регрессию, а также Ridge и  Lasso c параметрами по умолчанию и выведите их R2 на тестовой выборке.

In [6]:
from sklearn.linear_model import Ridge, Lasso, LinearRegression

In [7]:
names = ['Ridge', 'Lasso', 'Vanilla Linear']
models = [Ridge(), Lasso(), LinearRegression()]
for name, model in zip(names, models):
    model.fit(X_train, y_train)
    score = model.score(X_test, y_test)
    print(f'R^2 of {name} model: {score:.4f}')

R^2 of Ridge model: 0.6662
R^2 of Lasso model: 0.6671
R^2 of Vanilla Linear model: 0.6688


3. Для Ridge и Lasso подберите коэффициент регуляризации (используйте GridSearchCV, RidgeCV, LassoCV) в пределах от $10^{-5}$ до $10^5$ (по степеням 10). Посчитайте R2 на тестовой выборке по лучшим моделям и сравните с предыдущими результатами. Напишите как изменился результат.

In [8]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import RidgeCV, LassoCV

In [9]:
params = {'alpha': [10**n for n in range(-5, 6)]}
cv_models = []
cv_models.append(GridSearchCV(estimator=Ridge(), param_grid=params))
cv_models.append(GridSearchCV(estimator=Lasso(), param_grid=params))
cv_models.append(RidgeCV(alphas=params['alpha']))
cv_models.append(LassoCV(alphas=params['alpha']))

In [10]:
cv_names = ['Grid Search Ridge', 'Grid Search Lasso', 'Ridge CV', 'Lasso CV']
for name, model in zip(cv_names, cv_models):
    model.fit(X_train, y_train)
    score = model.score(X_test, y_test)
    print(f'R^2 of {name} model: {score:.4f}')

R^2 of Grid Search Ridge model: 0.6688
R^2 of Grid Search Lasso model: 0.6688
R^2 of Ridge CV model: 0.6688
R^2 of Lasso CV model: 0.6688


### Вывод:
R^2 не поменялся от регуляризации и остался на уровне обычной линейной регрессии (МНК)

4. Проведите масштабирование выборки (используйте Pipeline, StandardScaler, MinMaxScaler), посчитайте R2 и сравните с предыдущими результатами. Напишите как изменился результат

In [11]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, MinMaxScaler

In [12]:
models = [Ridge(), Lasso(), LinearRegression()]
m_names = ['ridge', 'lasso', 'vanilla_linear']
scalers = [StandardScaler(), MinMaxScaler()]
s_names = ['standard', 'min_max']
for model in zip(m_names, models):
    for scaler in zip(s_names, scalers):
        pipe = Pipeline(steps=[scaler, model])
        pipe.fit(X_train, y_train)
        score = pipe.score(X_test, y_test)
        print(f'R^2 of {model[0]} model with {scaler[0]} scaler: {score:.4f}')

R^2 of ridge model with standard scaler: 0.6685
R^2 of ridge model with min_max scaler: 0.6764
R^2 of lasso model with standard scaler: 0.6239
R^2 of lasso model with min_max scaler: 0.2574
R^2 of vanilla_linear model with standard scaler: 0.6688
R^2 of vanilla_linear model with min_max scaler: 0.6688


### Вывод:
Модели со стандартными настройками кроме Lasso показали примерно такой же результат.
Lasso показал очень плохой результат с Min Max масштабированием.

5. Подберите коэффициент регуляризации для Ridge и Lasso на масштабированных данных, посчитайте R2 и сравните с предыдущими результатами. Напишите как изменился результат

In [13]:
models = [RidgeCV(alphas=params['alpha']), LassoCV(alphas=params['alpha'])]
m_names = ['ridge_cv', 'lasso_cv']
scalers = [StandardScaler(), MinMaxScaler()]
s_names = ['standard', 'min_max']
for model in zip(m_names, models):
    for scaler in zip(s_names, scalers):
        pipe = Pipeline(steps=[scaler, model])
        pipe.fit(X_train, y_train)
        score = pipe.score(X_test, y_test)
        print(f'R^2 of {model[0]} model with {scaler[0]} scaler: {score:.4f}')

R^2 of ridge_cv model with standard scaler: 0.6660
R^2 of ridge_cv model with min_max scaler: 0.6700
R^2 of lasso_cv model with standard scaler: 0.6688
R^2 of lasso_cv model with min_max scaler: 0.6688


### Вывод:
Благодаря регулиризации R^2 вернулся в исходный уровень в 0.66-0.67

6. Добавьте попарные произведения признаков и их квадраты (используйте PolynomialFeatures) на масштабированных признаках, посчитайте R2 и сравните с предыдущими результатами. Напишите как изменился результат

In [14]:
from sklearn.preprocessing import PolynomialFeatures
import warnings
warnings.filterwarnings('ignore')

In [15]:
models = [RidgeCV(alphas=params['alpha']), LassoCV(alphas=params['alpha'])]
m_names = ['ridge_cv', 'lasso_cv']
scalers = [StandardScaler(), MinMaxScaler()]
s_names = ['standard', 'min_max']
for model in zip(m_names, models):
    for scaler in zip(s_names, scalers):
        pipe = Pipeline(steps=[scaler, ('pol_feat', PolynomialFeatures(include_bias=False)), model])
        pipe.fit(X_train, y_train)
        score = pipe.score(X_test, y_test)
        print(f'R^2 of {model[0]} model with {scaler[0]} scaler : {score:.4f}')

R^2 of ridge_cv model with standard scaler : 0.8180
R^2 of ridge_cv model with min_max scaler : 0.8501
R^2 of lasso_cv model with standard scaler : 0.8122
R^2 of lasso_cv model with min_max scaler : 0.8391


### Вывод:
Использование попарных произведений признаков значительно увеличило коэффициент детерминации до максимума 0.85 для Ridge с масштабированием min_max

7. Подберите наилучшую модель (используйте Pipeline, GridSearchCV) подбирая тип регуляризации (L1, L2), коэффициент регуляризации, метод масштабирования и степень полинома в PolynomialFeatures. Выведите итоговые параметры и результат R2. Напишите как изменился R2 по сравнению с предыдущими экспериментами

In [16]:
steps = [('scaler', StandardScaler()),
         ('polynomial', PolynomialFeatures(include_bias=False)),
         ('model', Ridge())]
params = {'scaler': [StandardScaler(), MinMaxScaler()],
          'polynomial__degree': [2, 3, 4],
          'model': [Ridge(), Lasso()],
          'model__alpha': params['alpha']
         }
pipe = Pipeline(steps=steps)
search_model = GridSearchCV(pipe, param_grid=params, n_jobs=2)

In [17]:
search_model.fit(X_train, y_train)

GridSearchCV(estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('polynomial',
                                        PolynomialFeatures(include_bias=False)),
                                       ('model', Ridge())]),
             n_jobs=2,
             param_grid={'model': [Ridge(), Lasso(alpha=0.001)],
                         'model__alpha': [1e-05, 0.0001, 0.001, 0.01, 0.1, 1,
                                          10, 100, 1000, 10000, 100000],
                         'polynomial__degree': [2, 3, 4],
                         'scaler': [StandardScaler(), MinMaxScaler()]})

In [18]:
score = search_model.score(X_test, y_test)
best_params = search_model.best_params_

In [19]:
from pprint import pprint
print('Parameters of the best model:')
pprint(best_params)
print()
print(f'R^2 of the best model: {score:.4f}')

Parameters of the best model:
{'model': Lasso(alpha=0.001),
 'model__alpha': 0.001,
 'polynomial__degree': 3,
 'scaler': MinMaxScaler()}

R^2 of the best model: 0.8450


### Вывод:
Как видим лучшая модель имеет L1 регуляризацию с парметром $\alpha = 0.001$, степень полинома 3 и метод масштабирования Min Max.
Однако на тестовых данных модель показала чуть меньший коэффициент детерминации, чем модель L2 на масштабировании Min Max и степени полинома 2. (0.845 против 0.850 соответственно)

http://archive.ics.uci.edu/ml/datasets/Adult<br>
https://raw.githubusercontent.com/jbrownlee/Datasets/master/adult-all.csv

age: continuous.<br>
workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.<br>
fnlwgt: continuous.<br>
education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.<br>
education-num: continuous.<br>
marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.<br>
occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.<br>
relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.<br>
race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.<br>
sex: Female, Male.<br>
capital-gain: continuous.<br>
capital-loss: continuous.<br>
hours-per-week: continuous.<br>
native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

In [20]:
columns = ['age',
           'workclass',
           'fnlwgt',
           'education',
           'education-num',
           'marital-status',
           'occupation',
           'relationship',
           'race',
           'sex',
           'capital-gain',
           'capital-loss',
           'hours-per-week',
           'native-country',
           'year-salary'
          ]

In [21]:
data = pd.read_csv('adult-all.csv', header=None)
data.columns = columns
data.tail()

age     workclass  fnlwgt  education  education-num  \
48837   39       Private  215419  Bachelors             13   
48838   64             ?  321403    HS-grad              9   
48839   38       Private  374983  Bachelors             13   
48840   44       Private   83891  Bachelors             13   
48841   35  Self-emp-inc  182148  Bachelors             13   

           marital-status       occupation    relationship  \
48837            Divorced   Prof-specialty   Not-in-family   
48838             Widowed                ?  Other-relative   
48839  Married-civ-spouse   Prof-specialty         Husband   
48840            Divorced     Adm-clerical       Own-child   
48841  Married-civ-spouse  Exec-managerial         Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
48837               White  Female             0             0              36   
48838               Black    Male             0             0              40   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country year-salary  
48837  United-States       <=50K  
48838  United-States       <=50K  
48839  United-States       <=50K  
48840  United-States       <=50K  
48841  United-States        >50K

8. Разделите выборку на признаки и целевую переменную(колонка со зачениями {<=50K,>50K}). Замените целевую переменную на числовые значения.

In [22]:
X, y = data.iloc[:, :-1], data.iloc[:, -1]
y = pd.Series(np.where(y == '<=50K', -1, 1))

9. Выясните, присутствуют ли в данных пропуски. Заполните их самыми частыми значениями (используйте SimpleImputer)

In [23]:
print('yearly-salary', y.isnull().any(), sep='\t  ')
X.isnull().any()

yearly-salary	  False


age               False
workclass         False
fnlwgt            False
education         False
education-num     False
marital-status    False
occupation        False
relationship      False
race              False
sex               False
capital-gain      False
capital-loss      False
hours-per-week    False
native-country    False
dtype: bool

10. Выберите колонки с числовыми и категориальными переменными.

In [24]:
cat_cols, num_cols = [], []
for name, tp in X.dtypes.to_dict().items():
    if tp == 'int64':
        num_cols.append(name)
    else:
        cat_cols.append(name)

print('categorial:', cat_cols, sep='\n')
print('numerical:', num_cols, sep='\n')

categorial:
['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
numerical:
['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']


11. Создайте пайплайн по обработке колонок (используйте OneHotEncoder, MinMaxScaler).

In [25]:
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

In [26]:
cat_trans = OneHotEncoder()
num_trans = MinMaxScaler()
trans = ColumnTransformer([('num', num_trans, num_cols), ('cat', cat_trans, cat_cols)])

12. Посчитайте метрики accuracy и f1_score на предсказании только самого частого класса в целевой переменной.

In [27]:
accuracy = max(y.value_counts()) / sum(y.value_counts())
print(f'Accuracy: {accuracy:.4f}')
TP = max(y.value_counts())
FP = min(y.value_counts())
TN = min(y.value_counts())
FN = 0
precision = TP / (TP + FP)
recall = TP / (TP + FN)
f1 = 2 * precision * recall / (precision + recall)
print(f'f1 score: {f1:.4f}')

Accuracy: 0.7607
f1 score: 0.8641


13. Посчитайте cross_val_score по алгоритмам LogisticRegression, SVC, LinearSVC по метрикам accuracy и f1_score.
Напишите удалось ли превзойти предыдущий результат.

In [28]:
from sklearn.svm import SVC, LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, f1_score

In [29]:
models = [LogisticRegression(random_state=RANDOM_STATE),
          LinearSVC(random_state=RANDOM_STATE)]
for model in models:
    pipe = Pipeline(steps=[('tranform', trans), ('model', model)])
    cv_acc = cross_val_score(pipe, X, y, scoring='accuracy')
    cv_f1 = cross_val_score(pipe, X, y, scoring='f1')
    print('Model:', model, '', sep='\n')
    print('Accuracy:', cv_acc, '', sep='\n')
    print('F1 score:', cv_f1, '', sep='\n')

Model:
LogisticRegression(random_state=42)

Accuracy:
[0.84901218 0.84850036        nan 0.85145373 0.8499181 ]

F1 score:
[0.65088757 0.65127238        nan 0.65898942 0.65408211]

Model:
LinearSVC(random_state=42)

Accuracy:
[0.85065002 0.85167366        nan 0.85176085 0.8531941 ]

F1 score:
[0.65137395 0.65736581        nan 0.65589354 0.65986717]



14. Можно заметить что в данных присутствуют значения '?', замените их самыми частыми значениями (испольуйте SimpleImputer)

In [30]:
from sklearn.impute import SimpleImputer

In [31]:
null_killer = SimpleImputer(missing_values='?', strategy='most_frequent')
X = pd.DataFrame(null_killer.fit_transform(X),
                 columns=X.columns).astype(X.dtypes.to_dict())
X.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
dtype: object

In [32]:
for col in X.columns:
    print(X[col].unique())

[39 50 38 53 28 37 49 52 31 42 30 23 32 40 34 25 43 54 35 59 56 19 20 45
 22 48 21 24 57 44 41 29 18 47 46 36 79 27 67 33 76 17 55 61 70 64 71 68
 66 51 58 26 60 90 75 65 77 62 63 80 72 74 69 73 81 78 88 82 83 84 85 86
 87 89]
['State-gov' 'Self-emp-not-inc' 'Private' 'Federal-gov' 'Local-gov'
 'Self-emp-inc' 'Without-pay' 'Never-worked']
[ 77516  83311 215646 ... 173449  89686 350977]
['Bachelors' 'HS-grad' '11th' 'Masters' '9th' 'Some-college' 'Assoc-acdm'
 'Assoc-voc' '7th-8th' 'Doctorate' 'Prof-school' '5th-6th' '10th'
 '1st-4th' 'Preschool' '12th']
[13  9  7 14  5 10 12 11  4 16 15  3  6  2  1  8]
['Never-married' 'Married-civ-spouse' 'Divorced' 'Married-spouse-absent'
 'Separated' 'Married-AF-spouse' 'Widowed']
['Adm-clerical' 'Exec-managerial' 'Handlers-cleaners' 'Prof-specialty'
 'Other-service' 'Sales' 'Craft-repair' 'Transport-moving'
 'Farming-fishing' 'Machine-op-inspct' 'Tech-support' 'Protective-serv'
 'Armed-Forces' 'Priv-house-serv']
['Not-in-family' 'Husband' 'Wife' 'O

15. Посчитайте cross_val_score на новых данных. Напишите удалось ли улучшить результат.

16. Посчитайте cross_val_score, если просто удалить значения '?'. Напишите как изменился результат

 17. Посчитайте cross_val_score для RandomForestClassifier, GradientBoostingClassifier. Напишите как изменился результат и какой вывод можно из этого сделать.

18. Подберите наилучшую модель, подбирая методы обработки колонок - масштабирование признаков, кодирование признаков и заполнение пропусков. Параметры алгоритмов оставьте по умолчанию. Выведите итоговые параметры и результат accuracy и f1_score.